// hermit-js/src/js/core/detect.js

import dom from "./dom";

const vendors = [
    "Webkit",
    "Moz",
    "ms"
];

export default {

    init () {
        this._isTouch = ("ontouchstart" in window || "DocumentTouch" in window);
        this._isMobile = (/Android|BlackBerry|iPhone|iPad|iPod|IEMobile|Opera Mini/gi.test( window.navigator.userAgent ));
        this._prefix = this.getPrefix();

        // Touch support mode
        if ( this._isTouch ) {
            dom.html.addClass( "is-touchable" );

        // Mouse support mode
        } else {
            dom.html.addClass( "is-hoverable" );
        }
    },

    getPrefix () {
        let i = vendors.length;

        if ( document.body.style.transform === null ) {
            for ( i; i--; ) {
                if ( document.body.style[ `${vendors[ i ]}Transform` ] ) {
                    return vendors[ i ];
                }
            }
        }

        return "";
    },

    getPrefixed ( property ) {
        const camelProp = (property[ 0 ].toUpperCase() + property.slice( 1 ));

        return (this._prefix ? (this._prefix + camelProp) : property);
    },

    isMobile () {
        return this._isMobile;
    },

    isTouch () {
        return this._isTouch;
    },

    isDevice () {
        return (this._isTouch && this._isMobile);
    }
};
