// hermit-js/src/js/controllers/pageController.js

import twig from 'twig';

import * as core from './../core';
import log from './../core/log';
import merge from './../core/merge';

import templateController from './templateController';
import moduleController from './moduleController';

let templateCache = [];

class PageController {

	constructor(){

		this.defaults = {
			templates: {}
		}

		this.initialized = false;
	}

	init(opts){

		log('PageController :: Init');

		this.initialized = true;

		this.options = merge(this.defaults, opts);
		this.templates = this.options.templates;
	}

	config(opts){
		this.options = merge({}, this.options, opts);
	}

	renderPage(data, match){

		let section = match.params.section || 'home';
		let templateUrl = this.templates[ match.route ];

		// didn't find the page template
		if( !templateUrl ){
			log('warn', 'PageController :: No template specified for route', match.route);
			return;
		}

		let template = templateController.getTemplate(match.route, templateUrl)
			.then((template)=>{

				let html = template.render(data);

				core.dom.main.html(`<div class="page page--${ section }">${ html }</div>`);

				requestAnimationFrame(()=>{
					moduleController.refresh();
				});
			});
	}
}

export default new PageController();
